Fix nnq.dropout in vision mobilenetv3 pretrain model

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71438

Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

terrychenism committed:
Fix nnq.dropout in vision mobilenetv3 pretrain model

Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -1633,6 +1635,23 @@ def forward(self, x):
+        x = self.fc2(x)
+        return self.dequant(x)
+
+class ManualDropoutQATModel(torch.nn.Module):
```
we can remove the two linears in the this model I think

terrychenism replied:
remove linear op create an empty parameter list, optimizer would fail.

jerryzh168 replied:
optional nit: I see, can we move one of the linears then?

terrychenism replied:
fc2 removed

jerryzh168 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -1499,7 +1499,8 @@ def convert(self,
 @register_quant_pattern(torch.nn.AvgPool2d)
 @register_quant_pattern(torch.nn.AvgPool3d)
-@register_quant_pattern(torch.nn.Dropout)
+# TODO fix it when codebase is stable
+# @register_quant_pattern(torch.nn.Dropout)
```
we can just remove this I think

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

jerryzh168 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -1509,7 +1508,6 @@ def convert(self,
 @register_quant_pattern(torch.adaptive_avg_pool1d)
 @register_quant_pattern(torch.nn.functional.adaptive_avg_pool2d)
 @register_quant_pattern(torch.nn.functional.adaptive_avg_pool3d)
```
should we support functional dropout? removing it is breaking the BC as well

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qdropout.cpp:
```diff
@@ -0,0 +1,21 @@
+
+DEFINE_DISPATCH(qdropout_stub);
+
+Tensor quantized_dropout(
```
I think this op should take a `training` argument, like https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1279

jerryzh168 replied:
nit: output_scale and output_zero_point should be the last two arguments of this op, otherwise looks good

jerryzh168 replied:
for the other parameters it would be best to match the order of original dropout op as well

jerryzh168 replied:
actually what is scale and input_scale?

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

jerryzh168 reviewed (changes requested):
there is some problem with the quantized_dropout signature I think

jerryzh168 commented on torch/nn/quantized/functional.py:
```diff
@@ -493,6 +493,19 @@ def elu(input: Tensor, scale: float, zero_point: int, alpha: float = 1.) -> Tens
         raise ValueError("Input to 'quantized.elu' must be quantized!")
     return torch.ops.quantized.elu(input, scale, zero_point, alpha)
 
+def dropout(input: Tensor, scale: float, zero_point: int, p: float = 0.5, training: bool = False) -> Tensor:
```
this is probably not used, I think we can remove

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
looks great, thanks!

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

terrychenism committed:
Update on "Fix nnq.dropout in vision mobilenetv3 pretrain model"


Summary:
Fix issue https://github.com/pytorch/vision/issues/5198
skip observer for nn.dropout to load pretrain model

Test Plan:
python -c "import torchvision; torchvision.models.quantization.mobilenet_v3_large(pretrained=True, quantize=True)"

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33641707](https://our.internmc.facebook.com/intern/diff/D33641707)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33641707).

datumbox commented:
Thanks for the fix. I hope we can merge soon to fix our broken CI.
